File: backend/app/models/notification.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base


class NotificationType(StrEnum):
    POST_PUBLISHED = "POST_PUBLISHED"
    POST_FAILED = "POST_FAILED"
    POST_SCHEDULED = "POST_SCHEDULED"
    INBOX_MESSAGE = "INBOX_MESSAGE"
    AI_SUGGESTION = "AI_SUGGESTION"
    SOCIAL_ACCOUNT = "SOCIAL_ACCOUNT"
    ANALYTICS = "ANALYTICS"
    SYSTEM = "SYSTEM"


class NotificationSeverity(StrEnum):
    INFO = "INFO"
    SUCCESS = "SUCCESS"
    WARNING = "WARNING"
    ERROR = "ERROR"


class Notification(Base):
    __tablename__ = "notifications"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    type: Mapped[NotificationType] = mapped_column(
        Enum(NotificationType, name="notification_type", create_type=False),
        nullable=False,
        index=True,
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    severity: Mapped[NotificationSeverity] = mapped_column(
        Enum(NotificationSeverity, name="notification_severity", create_type=False),
        default=NotificationSeverity.INFO,
        nullable=False,
    )
    is_read: Mapped[bool] = mapped_column(Boolean, default=False, index=True, nullable=False)
    action_url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    entity_type: Mapped[str | None] = mapped_column(String(100), nullable=True)
    entity_id: Mapped[str | None] = mapped_column(String(100), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        index=True,
        nullable=False,
    )
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    user = relationship("User")

File: frontend/communityai/src/features/notifications/types/notification.ts
export type NotificationType =
  | 'POST_PUBLISHED'
  | 'POST_FAILED'
  | 'POST_SCHEDULED'
  | 'INBOX_MESSAGE'
  | 'AI_SUGGESTION'
  | 'SOCIAL_ACCOUNT'
  | 'ANALYTICS'
  | 'SYSTEM'

export type NotificationSeverity = 'INFO' | 'SUCCESS' | 'WARNING' | 'ERROR'

export interface Notification {
  id: number
  type: NotificationType
  title: string
  message: string
  severity: NotificationSeverity
  is_read: boolean
  action_url?: string | null
  entity_type?: string | null
  entity_id?: string | null
  created_at: string
  read_at?: string | null
}

export interface NotificationListResponse {
  items: Notification[]
  total: number
  unread_count: number
  page: number
  limit: number
}

export interface NotificationUnreadCountResponse {
  unread_count: number
}

export interface NotificationFilters {
  type?: NotificationType
  severity?: NotificationSeverity
  is_read?: boolean
  page?: number
  limit?: number
}